File: src/Flagrum.Core/Graphics/Containers/GraphicsBinary.cs
using System.Collections.Generic;
using System.IO;
using Flagrum.Core.Serialization;
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Graphics.Containers;

public class GraphicsBinary : BinaryReaderWriterBase
{
    public uint Version { get; set; }
    public uint Unknown { get; set; }
    public Dictionary<string, string> Dependencies { get; set; }
    public IList<ulong> Hashes { get; set; }

    public override void Read(Stream stream)
    {
        using var reader = new MessagePackReader(stream, true);

        Version = reader.Read<uint>();
        reader.DataVersion = Version;

        // This property only exists in Episode Duscae
        if (Version <= 20141115)
        {
            Unknown = reader.Read<uint>();
        }

        Dependencies = reader.Read<Dictionary<string, string>>();
        Hashes = reader.Read<List<ulong>>();
    }

    public override void Write(Stream stream)
    {
        using var writer = new MessagePackWriter(stream, true);
        writer.DataVersion = Version;

        writer.Write(Version);
        writer.Write(Dependencies);
        writer.Write(Hashes);
    }
}

File: src/Flagrum.Blender/core/graphics/graphicsbinary.py
from dataclasses import dataclass

from ..serialization.messagepack.reader import MessagePackReader
from ..serialization.messagepack.writer import MessagePackWriter


@dataclass(init=False)
class GraphicsBinary:
    version: int
    unknown: int
    dependencies: dict[str, str]
    hashes: list[int]

    def __init__(self):
        self.version = 20160705
        self.dependencies = {}
        self.hashes = []

    def read(self, reader: MessagePackReader):
        self.version = reader.read()
        reader.data_version = self.version

        if reader.data_version <= 20141115:
            self.unknown = reader.read()

        self.dependencies = reader.read()

        hash_count = reader.read()
        self.hashes = []
        for i in range(hash_count):
            self.hashes.append(reader.read())

    def write(self, writer: MessagePackWriter):
        writer.write(self.version)
        writer.write(self.dependencies)
        writer.write(self.hashes)
